packages/pages-shared/__tests__/asset-server/handler.test.ts

```diff
@@ -920,6 +886,102 @@ describe("asset-server handler", () => {
 			);
 		});
 	});
+
+	describe("redirects", () => {
+		test("it should redirect uri-encoded paths", async () => {
+			const { response, spies } = await getTestResponse({
+				request: "https://foo.com/some%20page",
+				metadata: createMetadataObjectWithRedirects([
+					{ from: "/some%20page", to: "/home", status: 301 },
+				]),
+			});
+
+			expect(spies.fetchAsset).toBe(0);
+			expect(spies.findAssetEntryForPath).toBe(0);
+			expect(spies.getAssetKey).toBe(0);
+			expect(spies.negotiateContent).toBe(0);
+			expect(response.status).toBe(301);
+			expect(response.headers.get("Location")).toBe("/home");
+		});
+
+		test("redirects to a query string same-origin", async () => {
+			const { response } = await getTestResponse({
+				request: "https://foo.com/bar",
+				metadata: createMetadataObjectWithRedirects([
+					{ from: "/bar", to: "/?test=abc", status: 301 },
+				]),
+			});
+
+			expect(response.status).toBe(301);
+			expect(response.headers.get("Location")).toBe("/?test=abc");
+		});
+
+		test("redirects to a query string cross-origin", async () => {
+			const { response } = await getTestResponse({
+				request: "https://foo.com/bar",
+				metadata: createMetadataObjectWithRedirects([
+					{ from: "/bar", to: "https://foobar.com/?test=abc", status: 301 },
+				]),
+			});
+
+			expect(response.status).toBe(301);
+			expect(response.headers.get("Location")).toBe(
+				"https://foobar.com/?test=abc"
+			);
+		});
+
+		test("redirects to hash component same-origin", async () => {
+			const { response } = await getTestResponse({
+				request: "https://foo.com/bar",
+				metadata: createMetadataObjectWithRedirects([
+					{ from: "/bar", to: "https://foo.com/##heading-7", status: 301 },
+				]),
+			});
+
+			expect(response.status).toBe(301);
+			expect(response.headers.get("Location")).toBe("/##heading-7");
+		});
+
+		test("redirects to hash component cross-origin", async () => {
+			const { response } = await getTestResponse({
+				request: "https://foo.com/bar",
+				metadata: createMetadataObjectWithRedirects([
+					{ from: "/bar", to: "https://foobar.com/##heading-7", status: 301 },
+				]),
+			});
+
+			expect(response.status).toBe(301);
+			expect(response.headers.get("Location")).toBe(
+				"https://foobar.com/##heading-7"
+			);
+		});
+
+		test("redirects to a query string and hash same-origin", async () => {
+			const { response } = await getTestResponse({
+				request: "https://foo.com/bar",
+				metadata: createMetadataObjectWithRedirects([
+					{ from: "/bar", to: "/?test=abc#def", status: 301 },
+				]),
+			});
+
+			expect(response.status).toBe(301);
+			expect(response.headers.get("Location")).toBe("/?test=abc#def");
+		});
```

Comment:
```suggestion
		test("redirects to a query string and hash same-origin", async () => {
			const { response } = await getTestResponse({
				request: "https://foo.com/bar",
				metadata: createMetadataObjectWithRedirects([
					{ from: "/bar", to: "/?test=abc#def", status: 301 },
				]),
			});

			expect(response.status).toBe(301);
			expect(response.headers.get("Location")).toBe("/?test=abc#def");
		});
		
		test("redirects to a incoming query string and hash same-origin", async () => {
			const { response } = await getTestResponse({
				request: "https://foo.com/bar?test=abc",
				metadata: createMetadataObjectWithRedirects([
					{ from: "/bar", to: "/#def", status: 301 },
				]),
			});

			expect(response.status).toBe(301);
			expect(response.headers.get("Location")).toBe("/?test=abc#def");
		});
```

I think that's going to fail with `/#def?test=abc`.

Reply:
But also, that's not the result of your change here. That's just existing behavior, I think...

Reply:
Yeah I looked at this and per rfc, query string must be first - https://www.rfc-editor.org/rfc/rfc3986#section-4.1
Everything after hash is treated as a hash so I think this is actually ok

```
test("redirects to a query string and hash against rfc", async () => {
			const { response } = await getTestResponse({
				request: "https://foo.com/bar",
				metadata: createMetadataObjectWithRedirects([
					{ from: "/bar", to: "https://foobar.com/#def?test=abc", status: 301 },
				]),
			});

			expect(response.status).toBe(301);
			expect(response.headers.get("Location")).toBe(
				"https://foobar.com/#def?test=abc"
			);
		});

Reply:
That test you have written there is totally valid behavior. It's not against the RFC because, like you say, it's being (correctly) treated as part of the hash.

Mine was exposing a separate bug where we're incorrectly appending `.search` on the end of `.href`. In the case of `.href` containing anchor (`.hash`) parts and there are incoming `.search` parts, then the link we construct will be wrong. The anchor won't match up to anything on that redirected page and the `.search` parts will be dropped.

But yeah, separate bug which will be unaffected by this PR, best I can tell.

Reply:
Oh yeah, that's been the behaviour for a long time... I was trying to test this but I got 404, it never actually hit the redirect and I stopped looking at that point

Should probably look into it at some point though